### ISSUE TITLE
RFC: Support for Permissions (Access Control)

### ISSUE THREAD

**Issue by Martinsos** [wasp-lang/wasp#584]
# Support for Permissions (Access Control) in Wasp

## Motivation

When developing web apps, developers almost always need to implement some kind of access control, where they authenticate users/agents and then ensure that they only read data or perform actions that they are authorized for, nothing more.

We want to offer them an integrated, full-stack way in Wasp to handle most common use cases for access control, so that they can do it in easy and secure manner, while following best practices.

## Requirements

1. Centralized, well-structured, easy to review access control logic.
   - Recommended by OWASP, specifically they recommend ABAC over RBAC.
2. Deny access by default.
   - Recommended by OWASP, idea is that you don't by accident provide access that shouldn't be provided.
   - We can't deny everything by default though, it would be impractical.
3. Minimize possibility of developer mistakes where too much access is given.
   - Use case 3.1.: Piece of Entity (database data) is private, while the rest is public (in given context), and developer by accident shares the whole Entity when they shouldn't.
   - Use case 3.2.: Developer by accident allows user to read/modify Entity they shouldn't have access to.
     
In theory, we could collapse all 3 requirements above into the third requirement, which is minimizing chance of mistakes, since both requirements (1) and (2) serve that purpose, but I still separated them to make the discussion more practical and specific.

## Examples of developer mistakes that result in too much access given

1. Wasp by accident makes Wasp Operation public, due to forgetting to check if user is authenticated.
   - Requirement (3).
   - This check we usually do by putting `if (!context.user) { throw new HttpError(403) }` as the first line in the Operation. It is easy to forget to do this, happened to me.
   

2. Developer by accident sends the hashed password of the user to the frontend.
   - Requirement (3.1.).
   - They want to send the user's profile, which is represented by User Entity, and forget to exclude that field.

3. Developer by accident sends private user's profile data to another user.
   - Requirement (3.1.).
   - They want to send the summarized version of user's profile, which is represented by User Entity, and forget to exclude / not include some private fields like address, age or similar.

4. Developer by accident allows user to delete any article, even if it was made by another user.
   - Requirement (3.2.).
   - For example, user deletes article created by another user, although he should be able to do that. Or he reads draft of an article made by another user, but he shouldn't have access to it.

**Comment by shayneczyzewski** [comment]
Overall, I think this is a very nice structure and plan, @Martinsos ! The requirements make sense, and the examples solidified their relevance. Here are a few thoughts on the next steps:

Under Requirement 2 you note that:

> - We can't deny everything by default though, it would be impractical.

However, isn't this the proposal for next step?

> 1. Let's implement solution (2) -> make all Wasp Operations private by default, allow user to set `authRequired: false` to make an Operation public.

- Just curious what you had in mind regarding the impracticality in the first note.

- Also, I wonder if we need go give people a global flag for backwards compatibility that we can phase out, to make it less onerous for existing apps to upgrade?
- Lastly, I know we will have requirements for the next steps, but is your thinking that all `action` and `query` will get `authRequired` field like `page` do now, but just default to `true` and there is some Express middleware check added? Do we forsee any confusion here with `page` using the same thing but with different defaults for each (and behaviours, but that may be expected)?

In thinking ahead, I do think something like Casbin plus an auth middleware leveraging it to protect API routes makes sense. The entity field part feels much fuzzier and harder. However, in both cases I think the trick will be finding the right place/way to declare it in Wasp. Some future sci-fi DSL may help in solution (3). Great job!

**Comment by Martinsos** [comment]
@shayneczyzewski thanks!

Deny everything by default -> by everything I meant pages in frontend, operations in backend, all entity fields, maybe something else. For example denying all entity fields by default would probably be impractical. Same goes for frontend pages. But operations, for them it could be ok.

Global flag -> yes that might be an option! Or we can make it a breaking change, right now Wasp apps are pretty small so I don't think it should be a problem to upgrade. We are in Alpha, so that is ok, we want to figure out good defaults, which we can't do if we try to ensure backward compatibility.

If I got it right, what you are saying is that different defaults on frontend vs server side might be confusing? Yes that is a good point. I still thought they make sense, but I do agree that they might feel a bit confusing! Not sure how to go about it. Make everything private by default, both Pages and Operations? Maybe that would be ok. Plus provide app-wide flags for setting default behaviour?

Rules framework + middleware for Operations -> yup, that does seem like the most clear direction. Sounds good, I can work more on some sci-fi DSL suggestions / pieces of code!

**Comment by Martinsos** [comment]
We had a nice brainstorming on this, here are some ideas/thoughts from @shayneczyzewski @sodic @matijaSos @maksim36ua :
1. RBAC is widely used, but if fine granularity is needed, it can become too limiting.
2. Usual mechanism is middleware that gets inserted to do access control logic (Ruby on Rails via class inheritance, Node (Express) via middleware functions, Python via decorators, ...).
3. Auth0 has good middleware story, it is easy to insert into existing project, we might want to look at them for inspiration.
4. Supabase has row-level based security, that might be interesting as a concept.
5. While we could attach middleware directly to specific Operation, we could also go the opposite direction and specify for middleware which Operations it affects -> attach it to a group of Operations. So we could map either operation -> middleware(s) or middleware -> operation(s).
6. "Crazy" idea: Wasp could, once it has enough knowledge, list/show all the routes/operations/pages/other resources that specific user/role can access! So it could give developer a good idea of who has access to what.
7. For adding access control to Entity fields, we could maybe add "`.select()`" statements to entities that we inject via context into Operations -> something in that direction, basically ensure they are limited in what they can do. This can also maybe be done with monkey patching prisma code. Not sure if we can do this with Prisma really, but worth investigating a bit.
8. Idea for possible next step is to focus on the "middleware" part of the story as a start, so point (2) from the Summary in the RFC, which is how to attach the access control logic to stuff it guards (Operations, Entities). We can do this before introducing point (1) from Summary in RFC, which is a way to centralize and structure the rules. To take this step further, we might want to look into general way to insert middleware for Operations / Entities, instead of specifically inserting access control middleware - I am not sure what is better yet and what could be the differences though.

**Comment by shayneczyzewski** [comment]
### Permissions recap

@Martinsos and I spent some time brainstorming the permissions problem from multiple angles over a few sessions, and we did some prototyping with CASL as well. To recap, the challenge was to provide developers with an easy to use way to define permissions at the API level (Operations) and the Entity level, and determine what should be in Wasp DSL vs. what should be in JS. We also explored different views of the tradeoffs involved, for example manual vs. semi-automated vs. fully automated approaches, defining checks vs. using checks in DSL vs. JS, and even when the checks happen (e.g., before attempting to fetch vs. after). There was a lot covered, and the stream of consciousness notes can be found [here](https://www.notion.so/wasp-lang/Permissions-middleware-brainstorming-5cae9a04d50f4cc6a07ed596d9be6317).

For Operation-level authorization, different frameworks use different approaches. Rails, for instance, would allow you to get a lot of mileage out of just modifying your base `ApplicationController` (but you can also use middleware). Express tends to favor middleware.

If we decide to use middleware to secure Operations, we have a few options:

1) Expose [Express middleware](https://expressjs.com/en/guide/using-middleware.html#middleware.application) directly
   1) Pros: trivial to expose, allows devs to manipulate the request/response for things like `Header` manipulation, custom logging/error handling, and even use 3rd party Express middleware (of which there is a lot)
   2) Cons: tied directly to Express, low-level and requires boilerplate for the perms use case (but maybe helpers alleviate this)
2) Expose a slightly higher-level Wasp-specific middleware
   1) Pros: can still do interesting things, no longer tied to Express
   2) Cons: need to decide on use cases supported and design it
3) Expose a Wasp-specific, permissions-focused middleware
   1) Pros: would be focused squarely on permissions, so we could add lots of niceties to make it easy to do
   2) Cons: permissions-only (so less useful in general), need to design it

Other middleware considerations include determining if any of it can be leveraged for other application needs, like custom HTTP endpoints, web sockets, different authentication methods, etc.

We then spent some time writing sci-fi code for option (3) and thinking about what the permission check function interface could be, and how you could associate them with Operations. Part of the struggle here is we needed to pull info from the request/context (e.g., mapping args to real things you can check against, like "can the user update this specific Task?"), which became tedious to specify. Beyond that, the overarching question we kept hitting when trying various approaches was "why doesn't the dev just manually put these checks into their Operation functions?" Were the benefits gained by making the check function to Operation mapping in Wasp DSL better than that status-quo alternative? That is still an open question.

We then tried to think about enriching the Entities themselves so we could gain some of the benefits of what some GraphQL authorization libraries provide. They can do things more ergonomically at the API level since you know what the exact query/return shape are, but this requires having access to an Operation schema (which we currently lack in Wasp). They also do things on the Entity definitions for [field-level resolver checks](https://www.prisma.io/blog/graphql-directive-permissions-authorization-made-easy-54c076b5368e), which is something we could consider for basic checks (like RBAC-style).

Last, we tried looking at libraries for inspiration. Specifically, we were intrigued by [CASL](https://casl.js.org/v5/en/) in that it provided centralized ABAC as well as a [Prisma-aware query helpers](https://casl.js.org/v5/en/package/casl-prisma), in order to see if centralizing the rules was a better starting point than the check functions themselves.

CASL does make it nice to define "abilities", so we could use it for both Operation checks and Entity checks. We tried to see what it felt like using vanilla CASL, and then trying to make it easier to use here: https://github.com/wasp-lang/wasp/tree/martin-shayne-casl-perms/examples/wauth

Part of this effort was trying to auto-wire in both the ability checks in Operations and the CASL Prisma `where` helpers into the Prisma queries automatically via a JS Proxy, under a `context.checked.Task` sort of namespace.  This would mean anything you do will check the abilities automatically. This looked promising, but the Entity proxy approach gets complicated fast because:

- CASL Prisma only supports `WhereInput` (used in places where you fetch multiple results and more complex operators like `AND/OR` are available, so things like `updateMany()`), but it does not support `WhereUniqueInput` (used by things like `update()`).
- There is no easy way how to handle queries over related Entities via `connect`, `upsert`s, or `create` with associations [like so](https://www.prisma.io/docs/concepts/components/prisma-client/crud#advanced-query-examples).


### My closing thoughts

My recommendation is we step back and ask if doing the checks in code, like they can now, is "good enough" for Beta. If not, I propose the next step is we focus on providing support for library-agnostic (e.g., no CASL baked in) authorization of Operations via middleware. I would vote we delay explicit support for authorization on Entities at this time, until more use cases can be discussed and potential drawbacks above can be addressed.

TLDR: I think middleware would give us the most bang for the buck given our knowledge at this time, and something users will probably want/need anyways. In future phases I think we could refine this experience, add more explicit support for Entities, and hopefully move more of this into Wasp DSL.

**Comment by Martinsos** [comment]
Makes sense to me!

Just to recap a bit, I would say we are looking at couple of main directions right now:
1. General middleware for Wasp Operations, that can then also be used for ACL/permissions logic.
2. We integrate CASL library with Wasp and allow devs to use CASL to define their access control rules (checks), while Wasp then based on some conventions / rules will make sure to apply those on correct places. Specifically, CASL rule could be tied to Operation or Entity, and Wasp will then run them when that Operation is called or when that Entity is accessed via Prisma. They could still define rules with custom targets in CASL and call any of the rules manually, btw. This sounds pretty cool, especially for Entities, and helps devs feel confident that they won't by mistake do a thing they shouldn't. But, the part with ensuring that Prisma calls obey the rules requires significant effort, so this is what makes it hard. We explored this whole direction here https://github.com/wasp-lang/wasp/tree/martin-shayne-casl-perms/examples/wauth , and I tried to more concretely explain this approach here: https://github.com/wasp-lang/wasp/blob/f0b2d0d1be9aba6ff731d48ca7596ac4e747418e/examples/wauth/ext/acl.js#L55-L169 .
3. Similar to (2), but automatic applying of checks/rules by Wasp would be based on some kind of schema introduced at the Operation level (what Operations return). This would give us more knowledge and therefore more power probably, but Operations Schema is a major undertaking that is not yet clear, so we can't push on this one right now.

I am excited about (2) because it sounds like it could actually work, and like it could provide decent value and be pretty cool and innovative! But, it does sound like decent amount of work + lot of holes to plug and it is not 100% clear that we can make it work well enough. So one hand I am excited to give it a try, but on the other hand it might consume more resources than we expect in order to get it working.

So with that in mind, I also agree it makes sense to purse some simpler approach for now, like middleware (1). But I do hope we would relatively soon do at least a bit more of a push in the (2) direction!

**Comment by Martinsos** [comment]
User saying they would love to have RBAC, similar to casbin: https://discord.com/channels/686873244791210014/867713251479519232/1248925089311756382 .
